### PR TITLE
More options for variational/approximate multitask models

### PR DESCRIPTION
**Pull request by gpleiss**
- [x]  Rename `MultitaskVariationalStrategy` to `IndependentMultitaskVariationalStrategy`
- [x]  Add `LMCVariationalStrategy`
- [x]  Add example

### CODE REVIEW

**Comment by gpleiss** [comment]
@martinjankowiak and @jacobrgardner - this is a little different than what we had on the DSPP branch. As it is, LMCVariationalStrategy can be used to create an ICM model (shared kernel functions across latent functions) or a semi-parametric latent factor model (different kernel functions for each latent function).

File: docs/source/variational.rst [hunk 1]
```diff
@@ -84,6 +77,27 @@ deterministic relationship between :math:`\mathbf f` and :math:`\mathbf u`.
    :members:
 
 
+Variational Strategies for Multi-Ouptut Functions
```
Comment (martinjankowiak) [inline docs/source/variational.rst:80]: typo

File: gpytorch/variational/lmc_variational_strategy.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    .. note::
+
+        The batch shape of the base :obj:`~gpytorch.variational.VariationalStrategy` does not
+        necessarily have to correspond to the batch shape of the underlying GP objects.
+
+        For example, if the base variational strategy has a batch shape of `[3]` (corresponding
+        to 3 latent functions), the GP kernel object could have a batch shape of `[3]` or no
+        batch shape. This would correspond to each of the latent functions having different kernels
+        or the same kernel, respecitvely.
```
Comment (martinjankowiak) [inline gpytorch/variational/lmc_variational_strategy.py:36]: typo: respecitvely

File: gpytorch/variational/lmc_variational_strategy.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    r"""
+    LMCVariationalStrategy is an implementation of the "Linear Model of Coregionalization"
+    for multitask GPs. This model assumes that there are :math:`Q` latent functions
+    :math:`\mathbf g(\cdot) = [g^{(1)}(\cdot), \ldots, g^{(q)}(\cdot)]`,
+    each of which is modelled by a GP.
+    The output functions (tasks) are linear combination of the latent functions:
+
+    .. math::
+
+        f_{\text{task } i}( \mathbf x) = \sum_{q=1}^Q \mathbf a^{(q)} g^{(q)} ( \mathbf x )
```
Comment (martinjankowiak) [inline gpytorch/variational/lmc_variational_strategy.py:21]: is this the equation you mean?

File: gpytorch/variational/lmc_variational_strategy.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        self.num_latents = num_latents
+        self.latent_dim = latent_dim
+
+        # Make the batch_shape
+        self.batch_shape = list(batch_shape)
+        del self.batch_shape[self.latent_dim]
+        self.batch_shape = torch.Size(self.batch_shape)
+
+        # LCM coefficients
+        raw_lmc_coefficients = torch.randn(*batch_shape, self.num_tasks).tril_()
```
Comment (martinjankowiak) [inline gpytorch/variational/lmc_variational_strategy.py:103]: why tril?

Comment (gpleiss) [reply gpytorch/variational/lmc_variational_strategy.py]: that was leftover code. i caught it once i started writing an example notebook 😬

File: gpytorch/variational/lmc_variational_strategy.py (hunk 3 above)
Comment (gpleiss) [reply gpytorch/variational/lmc_variational_strategy.py:21]: That's this equation from Alvarez (2012):

Screen Shot 2020-06-17 at 12 59 53 PM

It leads to Kronecker structure when the latent functions share the same covariance. I think this is right?

Comment (martinjankowiak) [reply gpytorch/variational/lmc_variational_strategy.py]: that equation is fine; maybe i'm misreading your notation but your `a` only has 1 index not 2?

File: gpytorch/variational/lmc_variational_strategy.py [hunk 5]
```diff
@@ -18,7 +18,7 @@ class LMCVariationalStrategy(_VariationalStrategy):
 
     .. math::
 
-        f_{\text{task } i}( \mathbf x) = \sum_{q=1}^Q \mathbf a^{(q)} g^{(q)} ( \mathbf x )
+        f_{\text{task } i}( \mathbf x) = \sum_{q=1}^Q a_i^{(q)} g^{(q)} ( \mathbf x )
```
Comment (martinjankowiak) [inline gpytorch/variational/lmc_variational_strategy.py:21]: oh now it makes sense 👍

File: gpytorch/variational/lmc_variational_strategy.py (hunk 3 above)
Comment (gpleiss) [reply gpytorch/variational/lmc_variational_strategy.py]: ah yeah i see; there should be an `i` index as well.

**Pull request merged**